modules/ROOT/pages/traversal-framework/traversal_framework_java_api.adoc

```diff
@@ -0,0 +1,353 @@
+        }
+    }
+}
+----
+
+Here is an example of a combined evaluator, which will return all paths of length 2 that also have an end node with Label `A`:
+[source, java]
+----
+TraversalDescription td;
+try ( Transaction tx = graphDb.beginTx() ) {
+     td = tx.traversalDescription()
+            .evaluator(Evaluators.atDepth( 2 ))
+            .evaluator(new LabelEvaluator(Label.label("A")));
+}
+
+td.traverse( startNode );
+----
+
+[[traversal-java-api-traverser]]
+== Traverser
+
+The `Traverser` object is the result of invoking `traverse()` on a `TraversalDescription`.
+It represents a traversal positioned in the graph, and a specification of the format of the result.
+The actual traversal is performed lazily each time the `next()`-method of the iterator of the `Traverser` is invoked.
+
+=== Example
+Here is an example of a Traversal with default values (Uniqueness: _NODE_GLOBAL_, Expander: _BOTH_, Branch Ordering: _PREORDER_DEPTH_FIRST_):
+[source, java]
+----
+TraversalDescription td;
+try ( Transaction tx = graphDb.beginTx() ) {
+     td = tx.traversalDescription();
+}
+
+td.traverse( startNode );
+----
+
+[[traversal-java-api-uniqueness]]
+== Uniqueness
+
+Sets the rules for how positions can be revisited during a traversal as stated in `Uniqueness`.
+The default is `NODE_GLOBAL`.
+
+The various uniqueness levels that can be used in are:
+
+* `NONE` -- Any position in the graph may be revisited.
+* `NODE_GLOBAL` -- No node in the entire graph may be visited more than once.
+This could potentially consume a lot of memory since it requires keeping an in-memory data structure remembering all the visited nodes.
+* `RELATIONSHIP_GLOBAL` -- no relationship in the entire graph may be visited more than once.
+Just like `NODE_GLOBAL` uniqueness, this could potentially use up a lot of memory.
+But since graphs typically have a larger number of relationships than nodes, the memory overhead of this uniqueness level could grow even quicker.
+* `NODE_PATH` -- A node may not occur previously in the path reaching up to it.
+* `RELATIONSHIP_PATH` -- A relationship may not occur previously in the path reaching up to it.
+* `NODE_RECENT` -- Similar to `NODE_GLOBAL` uniqueness in that there is a global collection of visited nodes each position is checked against.
+This uniqueness level does however have a cap on how much memory it may consume in the form of a collection that only contains the most recently visited nodes.
+The size of this collection can be specified by providing a number as the second argument to the TraversalDescription.uniqueness()-method along with the uniqueness level.
+* `RELATIONSHIP_RECENT` -- Works like `NODE_RECENT` uniqueness, but with relationships instead of nodes.
+
+=== Example
+Here is an example of a traversal using a built-in `Uniqueness` constraint:
+----
+TraversalDescription td;
+try ( Transaction tx = graphDb.beginTx() ) {
+     td = tx.traversalDescription();
+            .uniqueness( Uniqueness.RELATIONSHIP_GLOBAL )
```

Comment:
I won't leave this comment for every case where this is happening, but there are more, so this needs thorogough review. The `;` needs to happen after all settings API have been called: 

```
     td = tx.traversalDescription()
            .uniqueness( Uniqueness.RELATIONSHIP_GLOBAL );
```